cmd/influx_tools/parquet/exporter.go

```diff
@@ -0,0 +1,359 @@
+		return err
+	}
+
+	if len(groups) == 0 {
+		return nil
+	}
+
+	sort.Sort(meta.ShardGroupInfos(groups))
+	e.sourceGroups = groups
+	e.startDate = groups[0].StartTime
+	e.endDate = groups[len(groups)-1].EndTime
+
+	return nil
+}
+
+func (e *exporter) shardsGroupsByTimeRange(min, max time.Time) []meta.ShardGroupInfo {
+	groups := make([]meta.ShardGroupInfo, 0, len(e.sourceGroups))
+	for _, g := range e.sourceGroups {
+		if !g.Overlaps(min, max) {
+			continue
+		}
+		groups = append(groups, g)
+	}
+	return groups
+}
+
+func (e *exporter) GatherInfo() error {
+	e.measurements = newMeasurements()
+	for _, g := range e.targetGroups {
+		min, max := g.StartTime, g.EndTime
+		rs, err := e.read(min, max.Add(-1))
+		if err != nil || rs == nil {
+			return err
+		}
+		e.gatherSchema(min, max, e.m, rs)
+		rs.Close()
+	}
+	return nil
+}
+
+func (e *exporter) WriteTo() error {
+	for _, g := range e.targetGroups {
+		min, max := g.StartTime, g.EndTime
+		rs, err := e.read(min, max.Add(-1))
+		if err != nil || rs == nil {
+			return err
+		}
+		if err := e.writeBucket(min, max, e.m, rs); err != nil {
+			return err
+		}
+		rs.Close()
+		fmt.Println("..")
+	}
+	fmt.Println(".")
+	return nil
+}
+
+// Read creates a ResultSet that reads all points with a timestamp ts, such that start ≤ ts < end.
+func (e *exporter) read(min, max time.Time) (*storage.ResultSet, error) {
+	shards, err := e.getShards(min, max)
+	if err != nil {
+		return nil, err
+	}
+
+	req := storage.ReadRequest{
+		Database: e.db,
+		RP:       e.rp,
+		Shards:   shards,
+		Start:    min.UnixNano(),
+		End:      max.UnixNano(),
+	}
+
+	return e.store.Read(context.Background(), &req)
+}
+
+func (e *exporter) Close() error {
+	return e.tsdbStore.Close()
+}
+
+func (e *exporter) getShards(min, max time.Time) ([]*tsdb.Shard, error) {
+	groups := e.shardsGroupsByTimeRange(min, max)
+	var ids []uint64
+	for _, g := range groups {
+		for _, s := range g.Shards {
+			ids = append(ids, s.ID)
+		}
+	}
+
+	shards := e.tsdbStore.Shards(ids)
+	if len(shards) == len(ids) {
+		return shards, nil
+	}
+
+	return e.openStoreWithShardsIDs(ids)
+}
+
+func (e *exporter) openStoreWithShardsIDs(ids []uint64) ([]*tsdb.Shard, error) {
+	e.tsdbStore.Close()
+	e.tsdbStore.EngineOptions.ShardFilter = func(_, _ string, id uint64) bool {
+		for i := range ids {
+			if id == ids[i] {
+				return true
+			}
+		}
+		return false
+	}
+	if err := e.tsdbStore.Open(); err != nil {
+		return nil, err
+	}
+	return e.tsdbStore.Shards(ids), nil
+}
+
+func (e *exporter) gatherSchema(start, end time.Time, measurement string, rs *storage.ResultSet) {
+	fmt.Printf("gather schema start: %s, end: %s\n", start.Format(time.RFC3339), end.Format(time.RFC3339))
+
+	for rs.Next() {
+		measurementName := string(models.ParseName(rs.Name()))
+		if measurement != "" && measurement != measurementName {
+			continue
+		}
+
+		t := e.measurements.getTable(measurementName)
+		t.addTags(rs.Tags())
+		t.addField(rs.Field(), rs.FieldType())
+	}
+}
```

Comment:
I would recommend you use the following, more efficient approach to gather the schema in InfluxDB 1.x. As mentioned in [this comment](https://github.com/influxdata/influxdb/pull/25047#issuecomment-2264522992), I recommend you process and export each shard separately to 1 or more parquet field, as there can be no schema conflicts _within a single shard_.

Given that, you will also be able to gather the complete schema _very efficiently_ using existing indices. 

For example, the shards returned by your `getShards` function returns a slice of `*tsdb.Shard`. Using that, you can get both the exact set of tag keys for that shard, and the set of fields:

```go
	cond := influxql.MustParseExpr("_name = '<measurement name>'")
	shard := shards[0]
	tagKeys, err := e.tsdbStore.TagKeys(context.Background(), query.OpenAuthorizer, []uint64{shard.ID()}, cond)
	fields := shard.MeasurementFields([]byte("<measurement name>"))
```

You could merge together the full set of tag keys across all shards to ensure the Parquet schema tag keys are consistent, and also perform a check that the field keys are all consistent data types. I would recommend generating a warning of field conflicts, but I would suggest you still export the individual shards, as the field types will be consistent within the shard.

Reply:
Thank you very much for the feedback. I have refactored the code per your suggestion, now individual shards are exported and schema is retrieved using the code above. The export is now single-pass op therefore.
Unfortunately, I still get the same wrong result (incomplete output).

Reply:
@alespour looking [at the code](https://github.com/influxdata/influxdb/blob/3697f752a422e96c38512c7fb36e391b65c58465/cmd/influx_tools/parquet/exporter.go#L186), it seems like you do not create a schema if a measurement has no tags. Even though this might be rare, it's perfectly valid to have a series without _any_ tag... Do you by chance miss data in the output from measurements without tags?

Reply:
@srebhan I believe the database and measurement (telegraf, cpu / disk etc) I 'm using for testing does not contain tag-less data.

Reply:
I'm using telegraf to check parquet output like this:
`telegraf --config ./telegraf-parquet.conf --once`
```
[[inputs.file]]
  files = ["/tmp/parquet/table-*.parquet"]
  name_override = "cpu"
  data_format = "parquet"
  tag_columns = ["datacenter","hostname","os","rack","region","service","team"]
  timestamp_column = "time"
  timestamp_format = "unix_ns"

[[outputs.file]]
  ## Files to write to, "stdout" is a specially handled file.
  files = ["stdout"]
```